### PR TITLE
README additions

### PR DESCRIPTION
**Pull request by coderdan**
Updated the README to make sure insecure code doesn't get implemented by copy-pasta.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -179,6 +177,23 @@ class User {
...
+is GCM ([Galois Counter Mode](https://en.wikipedia.org/wiki/Galois/Counter_Mode)).
+
+GCM provides both secrecy and authenticity and can generally use CPU acceleration where available.
+
+### Should I hardcode the IV?
+
+No. Don't ever do this.
+It will break the encryption and is vulnerable to a "repeated nonce" attack.
+
+If you don't provide an IV, the library will randomly generate a secure one for you.
```
Comment (generalpiston) [inline README.md:194]: Unfortunately, I think GCM or CTR modes require the set of IVs to be perfectly disjoint. There can be no overlap, basically.

The code currently generates IVs randomly if not provided. It doesn't guarantee uniqueness.

For GCM mode, perhaps we should allow `options.iv` to be a function where library users can provide a function that guarantees uniqueness.

File: README.md [hunk 2]
```diff
@@ -179,6 +177,23 @@ class User {
 
 Queries that transform the encrypted column wont work because transformers and subscribers operate outside of the DBMS.
 
+### What alogorithm should I use?
+
+Unless you need to maintain compatibility with an older system (or you know exactly what you're doing),
+you should use "aes-256-gcm" for the mode.
```
Comment (generalpiston) [inline README.md:183]: Unfortunately, I think GCM or CTR modes require the set of IVs to be unique. There can be no overlap, basically.

The code currently generates IVs randomly if not provided. It doesn't guarantee uniqueness.

For GCM mode, perhaps we should allow `options.iv` to be a function where library users can provide a function that guarantees uniqueness.

Until then, CBC might be safer.

File: README.md (hunk 1 above)
Comment (generalpiston) [inline README.md:194]: Maybe "In production, hardcoding an IV is not secure. See repeated nonce attack." And link to https://www.elttam.com/blog/key-recovery-attacks-on-gcm/.

File: README.md [hunk 3]
```diff
@@ -179,6 +177,23 @@ class User {
...
+This means that the encryption keys are are 256 bits (32-bytes) long and that the mode of operation
+is GCM ([Galois Counter Mode](https://en.wikipedia.org/wiki/Galois/Counter_Mode)).
+
+GCM provides both secrecy and authenticity and can generally use CPU acceleration where available.
+
+### Should I hardcode the IV?
+
+No. Don't ever do this.
```
Comment (generalpiston) [inline README.md:191]: Strong words... in a test case I think using the same IV is fine.

File: README.md [hunk 4]
```diff
@@ -179,6 +177,23 @@ class User {
 
 Queries that transform the encrypted column wont work because transformers and subscribers operate outside of the DBMS.
 
+### What alogorithm should I use?
+
+Unless you need to maintain compatibility with an older system (or you know exactly what you're doing),
```
Comment (generalpiston) [inline README.md:182]: I think these algorithms have different properties that have different applications and requirements. We can give advice like "aes-256-gcm works well for encryption at rest" and point people to more information. But I don't think we should give off a tone where all of this stuff seems beyond them.

So how about this:

```
Which algorithm should I use?

aes-256-gcm works well for encryption at rest.

For a break down of algorithms, see https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb.
```

File: README.md (hunk 2 above)
Comment (coderdan) [reply README.md:183]: CBC requires unique IVs, too. All modes of operation for AES to (with the possible exception of SIV which isn't available in Node anyway).

If the IVs are 16-bytes long and generated with a Cryptographically Secure Pseudo Random Number Generator (CSPRNG) then that is perfectly fine. The `randomBytes` function (that is used internally now) meets this requirement.

Providing a function to generate the IV would work but puts a lot of responsibility on the developer and could make mistakes more likely.

File: README.md (hunk 3 above)
Comment (coderdan) [reply README.md:191]: Strong words indeed but for good reason! How about we say "No. Don't ever do this in production." ?

File: README.md (hunk 4 above)
Comment (coderdan) [reply README.md:182]: I have no problem in pointing people to educational resources. Having said that, its probably worth mentioning that unless you understand this stuff well that GCM is generally the best choice.

CBC does not provide authenticity and so is vulnerable to chosen ciphertext attacks.

**Comment by generalpiston** [comment]
Thanks @coderdan for your PR. Great work!

**Pull request merged**